### PR TITLE
Make sure the node's IP addresses is set in noProxy

### PR DESCRIPTION
**Pull request by votdev**
grafik

**Problem:**
If you configure httpProxy and httpsProxy, you must also put Harvester node's CIDR into noProxy, otherwise, the Harvester cluster will be broken.

**Solution:**
Add a webhook to validate the user input; it should reject if the node IP is not included in the no_proxy CIDR range.

**Related Issue:**
https://github.com/harvester/harvester/issues/4282

**Test plan:**
The test plan assumes that several hosts are available in the cluster. For the following test cases, the hosts with the following IP addresses are involved:

- 192.168.0.30
- 192.168.0.31
- 192.168.0.32

## Case 1:
1. Go to `Advanced` > `Settings` and then select `Edit Settings` for `http-proxy`.
2. Press `Save`.
3. The error message `noProxy should contain the node's IP addresses or CIDR. The node(s) 192.168.0.30, 192.168.0.31, 192.168.0.32 are not covered.` is displayed.

grafik

## Case 2:
1. Go to `Advanced` > `Settings` and then select `Edit Settings` for `http-proxy`.
2. Use `192.168.0.0/27` in the `no-proxy` form field.
3. Press `Save`.
4. The error message  `noProxy should contain the node's IP addresses or CIDR. The node(s) 192.168.0.32 are not covered.` is displayed.

## Case 3:
1. Go to `Advanced` > `Settings` and then select `Edit Settings` for `http-proxy`.
2. Use `192.168.0.0/24` in the `no-proxy` form field.
3. Press `Save`.
4. The setting is saved and you'll be redirected to the `Settings` page.

## Case 4:
1. Go to `Advanced` > `Settings` and then select `Edit Settings` for `http-proxy`.
2. Use `192.168.0.30, 192.168.0.31, 192.168.0.32` in the `no-proxy` form field.
3. Press `Save`.
4. The setting is saved and you'll be redirected to the `Settings` page.

### CODE REVIEW

**Comment by Vicente-Cheng** [comment]
Hi @votdev, 
Could you move the common function (e.g. mpa/slice) to https://github.com/harvester/go-common?
That we could reuse this in various components!

Thanks!

**Comment by votdev** [comment]
> Hi @votdev, Could you move the common function (e.g. mpa/slice) to [harvester/go-common](https://github.com/harvester/go-common)? That we could reuse this in various components!
> 
> Thanks!

See https://github.com/harvester/go-common/pull/14

File: pkg/webhook/resources/setting/validator.go [hunk 1]
```diff
@@ -213,19 +223,81 @@ func validateDeleteSetting(oldObj runtime.Object) error {
...
+	// Make sure the node's IP addresses is set in 'noProxy'. These IP
+	// addresses can be specified individually or via CIDR address.
+	nodes, err := v.nodeCache.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	err = validateNoProxy(httpProxyConfig.NoProxy, nodes)
+	if err != nil {
+		return err
+	}
```
Comment (Yu-Jack) [inline pkg/webhook/resources/setting/validator.go]: Should they use `werror.NewInvalidError` to create validation error as well?

Comment (votdev) [reply pkg/webhook/resources/setting/validator.go]: > Should they use `werror.NewInvalidError` to create validation error as well?

The function `validateNoProxy` returns `NewInvalidError` errors. But you are right for the `nodes, err := v.nodeCache.List(labels.Everything())` error case.

**Review by FrankYang0529 — APPROVED** [review]
LGTM. Thanks for the PR.

**Review by Yu-Jack — APPROVED** [review]
LGTM, thanks!

**Comment by mingshuoqiu** [comment]
LGTM.

**Comment by votdev** [comment]
@Mergifyio backport v1.2 v1.3

**Pull request merged**